libs/base/Data/Colist1.idr

```diff
@@ -0,0 +1,205 @@
+module Data.Colist1
+
+import Data.Colist
+import Data.List1
+import Data.Nat
+import public Data.Zippable
+
+%default total
+
+||| A possibly infinite, non-empty list.
+public export
+data Colist1 : (a : Type) -> Type where
+  (:::) : a -> Colist a -> Colist1 a
+
+--------------------------------------------------------------------------------
+--          Creating Colist1
+--------------------------------------------------------------------------------
+
+||| Convert a `List1` to a `Colist1`.
+public export
+fromList1 : List1 a -> Colist1 a
+fromList1 (h ::: t) = h ::: fromList t
+
+||| Convert a stream to a `Colist1`.
+public export
+fromStream : Stream a -> Colist1 a
+fromStream (x :: xs) = x ::: fromStream xs
+
+||| Try to convert a `Colist` to a `Colist1`. Returns `Nothing`
+||| the given `Colist` is empty.
+public export
+fromColist : Colist a -> Maybe (Colist1 a)
+fromColist Nil       = Nothing
+fromColist (x :: xs) = Just (x ::: xs)
+
+||| Try to convert a list to a `Colist1`. Returns `Nothing`
+||| the given list is empty.
+public export
+fromList : List a -> Maybe (Colist1 a)
+fromList = fromColist . fromList
+
+||| Create a `Colist1` of only a single element.
+public export
+singleton : a -> Colist1 a
+singleton a = a ::: Nil
+
+||| An infinite `Colist1` of repetitions of the same element.
+public export
+repeat : a -> Colist1 a
+repeat v = v ::: repeat v
+
+||| Create a `Colist1` of `n` replications of the given element.
+public export
+replicate : (n : Nat) -> {auto 0 prf : IsSucc n} -> a -> Colist1 a
+replicate 0     _ impossible
+replicate (S k) x = x ::: replicate k x
+
+||| Produce a `Colist1` by repeating a sequence
+public export
+cycle : List1 a -> Colist1 a
+cycle (h ::: [])          = repeat h
+cycle (h ::: t@(y :: ys)) = h ::: run y ys
+  where run : a -> List a -> Colist a
+        run v []        = v :: run h t
+        run v (y :: ys) = v :: run y ys
```

Comment:
```suggestion
cycle (x ::: xs) = run x xs
  where run : a -> List a -> Colist a
        run v []        = v :: run x xs
        run v (y :: ys) = v :: run y ys
```

Reply:
I don't think this works, does it? run should now have return type `Colist1 a` instead of `Colist a`. A shorter version
that would work would be

```
cycle (h ::: t) = h ::: cycle (t <+> [h])
```

That would be short and clean.

Reply:
Changed the implementation to `cycle (h ::: t) = h ::: cycle (t ++ [h])`.

Reply:
Good point. It's pretty annoying we don't have a type level notion of
"being guarded" because we really ought to be able to write this like
so:

```idris
cycle1 : List1 a -> Colist1 a
cycle1 xxs = xxs `lappend1` cycle1 xxs
```

where `lappend1 : List1 a -> GUARDED (Colist1 a) -> Colist1 a`

Reply:
Something like this "GUARDED" keyword would indeed be quite useful. I'm at the moment experimenting with Cotrees (possibly infinite rose trees) and writing total functions often means repeating quite a bit of code, since it's for instance not possible to use a recursive call in Functor `map`.

Anyway, if my experiments lead to some useful results and if there is some interest in having Tree and Cotree in base or contrib, I'll be happy to add those as well.